tests/bin/sessions.test.ts

```diff
@@ -1,215 +1,321 @@
+      exitCode,
+      stderr,
+      new clientErrors.ErrorClientAuthDenied(),
+    );
+    // Password set
+    ({ exitCode, stderr } = await testBinUtils.pkStdio(
+      ['agent', 'status'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+        PK_PASSWORD: 'invalid',
+        PK_TOKEN: undefined,
+      },
+      global.binAgentDir,
+    ));
+    testBinUtils.expectProcessError(
+      exitCode,
+      stderr,
+      new clientErrors.ErrorClientAuthDenied(),
+    );
+    // Token set
+    ({ exitCode, stderr } = await testBinUtils.pkStdio(
+      ['agent', 'status'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+        PK_PASSWORD: undefined,
+        PK_TOKEN: 'token',
+      },
+      global.binAgentDir,
+    ));
+    testBinUtils.expectProcessError(
+      exitCode,
+      stderr,
+      new clientErrors.ErrorClientAuthDenied(),
+    );
+  });
+  test('password can be used to authenticate attended calls', async () => {
+    const password = global.binAgentPassword;
+    mockedPrompts.mockClear();
+    mockedPrompts.mockImplementation(async (_opts: any) => {
+      return { password };
     });
+    const { exitCode, stdout } = await testBinUtils.pkStdio(
+      ['agent', 'status', '--verbose'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+      },
+      global.binAgentDir,
+    );
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    // Prompted for password 1 time
+    expect(mockedPrompts.mock.calls.length).toBe(1);
+    mockedPrompts.mockClear();
+  });
+  test('re-prompt for password if unable to authenticate call', async () => {
+    const validPassword = global.binAgentPassword;
+    const invalidPassword = 'invalid';
+    mockedPrompts.mockClear();
+    mockedPrompts
+      .mockResolvedValueOnce({ password: invalidPassword })
+      .mockResolvedValue({ password: validPassword });
+    const { exitCode, stdout } = await testBinUtils.pkStdio(
+      ['agent', 'status', '--verbose'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+      },
+      global.binAgentDir,
+    );
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    // Prompted for password 2 times
+    expect(mockedPrompts.mock.calls.length).toBe(2);
+    mockedPrompts.mockClear();
+  });
+  test('will not prompt for password reauthentication on generic error', async () => {
+    const validPassword = global.binAgentPassword;
+    const invalidPassword = 'invalid';
+    mockedPrompts.mockClear();
+    mockedPrompts
+      .mockResolvedValueOnce({ password: invalidPassword })
+      .mockResolvedValue({ password: validPassword });
+    const { exitCode } = await testBinUtils.pkStdio(
+      ['identities', 'search', 'InvalidProvider'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+      },
+      global.binAgentDir,
+    );
+    expect(exitCode).not.toBe(0);
+    // Prompted for password 2 times
+    expect(mockedPrompts.mock.calls.length).toBe(2);
+    mockedPrompts.mockClear();
```

Comment:
Potentially a misunderstanding on my part, but what's this test achieving? As far as I understand, we provide some incorrect call, and:
1. Are prompted for a password, with which we give an incorrect password
2. Are prompted again for a password, with which we give the correct password (so no more password prompting would occur)
3. The call goes through, but it's an incorrect call, so throws some kind of error

However, the test description says "will not prompt for password reauthentication on generic error" - why would it prompt for a password reauthentication, given that we gave the correct password in step 2?

Reply:
Yes this is right. It should prompt for password on generic error. So a correct test would be to provide the password on the right call in the beginning and expect an error. But how are you simulating an error here? Through deliberate mocking?

Reply:
This might already be well tested in the `utils.retryAuth.test.ts`.

Reply:
Yeah the tests here are pretty much just copied and pasted from `utils.retryAuth.test.ts`. I added them after what you said in our meeting yesterday.

Reply:
> But how are you simulating an error here? Through deliberate mocking?

The error is thrown by `identities search` with an invalid provider. I only mocked the prompts library for these tests. So it's just checking that the error will be thrown as expected rather than catching it and prompting for the password again (as if it were an authentication error).

Reply:
I've removed this test for now. It's too brittle. And the behaviour is well tested by `utils.retryAuth.test.ts`.